src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs

```diff
@@ -0,0 +1,822 @@
+        }
+
+        internal void EndProcessAuthentication(IAsyncResult result)
+        {
+            if (result == null)
+            {
+                throw new ArgumentNullException("asyncResult");
+            }
+
+            LazyAsyncResult lazyResult = result as LazyAsyncResult;
+            if (lazyResult == null)
+            {
+                throw new ArgumentException(SR.Format(SR.net_io_async_result, result.GetType().FullName), "asyncResult");
+            }
+
+            if (Interlocked.Exchange(ref _nestedAuth, 0) == 0)
+            {
+                throw new InvalidOperationException(SR.Format(SR.net_io_invalidendcall, "EndAuthenticate"));
+            }
+
+            // No "artificial" timeouts implemented so far, InnerStream controls that.
+            lazyResult.InternalWaitForCompletion();
+
+            Exception e = lazyResult.Result as Exception;
+
+            if (e != null)
+            {
+                // Round-trip it through the SetException().
+                e = SetException(e);
+                throw e;
+            }
+        }
+
+        private bool CheckSpn()
+        {
+            if (_context.IsKerberos)
+            {
+                return true;
+            }
+
+            if (_extendedProtectionPolicy.PolicyEnforcement == PolicyEnforcement.Never ||
+                    _extendedProtectionPolicy.CustomServiceNames == null)
+            {
+                return true;
+            }
+
+            string clientSpn = _context.ClientSpecifiedSpn;
+
+            if (String.IsNullOrEmpty(clientSpn))
+            {
+                if (_extendedProtectionPolicy.PolicyEnforcement == PolicyEnforcement.WhenSupported)
+                {
+                    return true;
+                }
+            }
+            else
+            {
+                return _extendedProtectionPolicy.CustomServiceNames.Contains(clientSpn);
+            }
+
+            return false;
+        }
+
+        //
+        // Client side starts here, but server also loops through this method.
+        //
+        private void StartSendBlob(byte[] message, LazyAsyncResult lazyResult)
+        {
+            Exception exception = null;
+            if (message != s_emptyMessage)
+            {
+                message = GetOutgoingBlob(message, ref exception);
+            }
+
+            if (exception != null)
+            {
+                // Signal remote side on a failed attempt.
+                StartSendAuthResetSignal(lazyResult, message, exception);
+                return;
+            }
+
+            if (HandshakeComplete)
+            {
+                if (_context.IsServer && !CheckSpn())
+                {
+                    exception = new AuthenticationException(SR.net_auth_bad_client_creds_or_target_mismatch);
+                    int statusCode = ERROR_TRUST_FAILURE;
+                    message = new byte[8];  //sizeof(long)
+
+                    for (int i = message.Length - 1; i >= 0; --i)
```

Comment:
I see three copies of this currently on my screen.  Perhaps factor it out to a method?